src/main/java/be/seeseemelk/mockbukkit/scheduler/BukkitSchedulerMock.java

```diff
@@ -281,67 +304,91 @@ public boolean isQueued(int taskId)
 	}
 
 	@Override
-	public List<BukkitWorker> getActiveWorkers()
+	public @NotNull List<BukkitWorker> getActiveWorkers()
 	{
 		// TODO Auto-generated method stub
 		throw new UnimplementedOperationException();
 	}
 
 	@Override
-	public List<BukkitTask> getPendingTasks()
+	public @NotNull List<BukkitTask> getPendingTasks()
 	{
 		// TODO Auto-generated method stub
 		throw new UnimplementedOperationException();
 	}
 
 	@Override
-	public BukkitTask runTaskAsynchronously(Plugin plugin, Runnable task)
+	public @NotNull BukkitTask runTaskAsynchronously(@NotNull Plugin plugin, @NotNull Runnable task)
 	{
+
 		ScheduledTask scheduledTask = new ScheduledTask(id++, plugin, false, currentTick, new AsyncRunnable(task));
+		checkForShutDown(scheduledTask);
 		asyncTasksRunning.incrementAndGet();
 		pool.execute(scheduledTask.getRunnable());
 		return scheduledTask;
 	}
 
 	@Override
-	public BukkitTask runTaskAsynchronously(Plugin plugin, BukkitRunnable task)
+	public @NotNull BukkitTask runTaskAsynchronously(@NotNull Plugin plugin, @NotNull BukkitRunnable task)
 	{
 		return runTaskAsynchronously(plugin, (Runnable) task);
 	}
 
 	@Override
-	public BukkitTask runTaskLater(Plugin plugin, BukkitRunnable task, long delay)
+	public @NotNull BukkitTask runTaskLater(@NotNull Plugin plugin, @NotNull BukkitRunnable task, long delay)
 	{
 		return runTaskLater(plugin, (Runnable) task, delay);
 	}
 
 	@Override
-	public BukkitTask runTaskLaterAsynchronously(Plugin plugin, Runnable task, long delay)
+	public @NotNull BukkitTask runTaskLaterAsynchronously(@NotNull Plugin plugin, @NotNull Runnable task, long delay)
 	{
 		ScheduledTask scheduledTask = new ScheduledTask(id++, plugin, false, currentTick + delay,
 		        new AsyncRunnable(task));
-		tasks.add(scheduledTask);
+		checkForShutDown(scheduledTask);
 		asyncTasksQueued++;
+		tasks.add(scheduledTask);
 		return scheduledTask;
 	}
 
 	@Override
-	public BukkitTask runTaskLaterAsynchronously(Plugin plugin, BukkitRunnable task, long delay)
+	public @NotNull BukkitTask runTaskLaterAsynchronously(@NotNull Plugin plugin, @NotNull BukkitRunnable task, long delay)
 	{
 		return runTaskLaterAsynchronously(plugin, (Runnable) task, delay);
 	}
 
 	@Override
-	public BukkitTask runTaskTimerAsynchronously(Plugin plugin, Runnable task, long delay, long period)
+	public @NotNull BukkitTask runTaskTimerAsynchronously(@NotNull Plugin plugin, @NotNull Runnable task, long delay, long period)
 	{
 		RepeatingTask scheduledTask = new RepeatingTask(id++, plugin, false, currentTick + delay, period,
 		        new AsyncRunnable(task));
+		checkForShutDown(scheduledTask);
 		tasks.add(scheduledTask);
 		return scheduledTask;
 	}
 
+	private void checkForShutDown(ScheduledTask task)
```

Comment:
I reckon we could add an annotation here as well.
```suggestion
	private void checkForShutDown(@NotNull ScheduledTask task)
```

Reply:
Yeah I guess because it was private I didn't really get overly concerned

Reply:
Like I said, not a big deal.
With private methods we have full control over what goes into them anyway, it was just something I wanted to note, even if it is nitpicky.

Reply:
I personally think `check*` is **always** a bad name for a method. A `check*` method presumbly does some calculation, and then what? Does it throw away the result? Does it return it? Does it perform some task? Execute some callbacks? Something else?

This method would perhaps be called more appropriately `assertSchedulable(Task)`.